src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,700 @@
+        """
+        Convert full block diagonal matrix into list representation
+
+        :param mat; The full block diagonal matrix including the zero elements
+        ofnon-diagonal blocks.
+        :param blk_partition: The matrix block partition in the form of a
+        K-element list storing all shapes of K diagonal matrix blocks,
+        where `blk_partition[i]` corresponds to the shape (number rows
+        and columns) of the `i` diagonal matrix block.
+
+        :return: The BlkDiagMatrix matrix
+        """
+
+        # TODO: maybe can improve this implementation
+
+        A = BlkDiagMatrix(len(blk_partition), dtype=dtype)
+
+        rows = blk_partition[:, 0]
+        cols = blk_partition[:, 1]
+        cellarray = Cell2D(rows, cols, dtype=mat.dtype)
+        blk_diag = cellarray.mat2blk_diag(mat, rows, cols)
+        A.data = BlkDiagMatrix.from_blk_diag(blk_diag)
+        return A
+
+    def solve(self, Y):
+        """
+        Solve a linear system involving a block diagonal matrix
+
+        :param Y: The right-hand side in the linear system.  May be a matrix
+        consisting of  coefficient vectors, in which case each column is
+        solved for separately.
+
+        :return: The result of solving the linear system formed by the matrix.
+        """
+
+        rows = np.array([np.size(mat_a, 0) for mat_a in self])
+        if sum(rows) != np.size(Y, 0):
+            raise RuntimeError('Sizes of `self` and `Y` are not compatible.')
+
+        vector = False
+        if np.ndim(Y) == 1:
+            Y = Y[:, np.newaxis]
+            vector = True
+
+        cols = np.array([np.size(Y, 1)])
+        cellarray = Cell2D(rows, cols, dtype=Y.dtype)
+        Y = cellarray.mat2cell(Y, rows, cols)
+        X = []
+        for i in range(0, self.nblocks):
+            X.append(solve(self[i], Y[i]))
+        X = np.concatenate(X, axis=0)
+
+        if vector:
+            X = X[:, 0]
+
+        return X
+
+    def apply(self, X):
+        """
+        Define the apply option of a block diagonal matrix with a matrix of
+        coefficient vectors.
+
+        :param X: The coefficient matrix with each col is a coefficient vector
+
+        :return: A matrix with new coefficient vectors
+        """
+        cols = np.array([np.size(mat, 1) for mat in self])
+
+        if np.sum(cols) != np.size(X, 0):
+            raise RuntimeError(
+                'Sizes of matrix `self` and `X` are not compatible.')
+
+        vector = False
+        if np.ndim(X) == 1:
+            X = X[:, np.newaxis]
+            vector = True
+
+        rows = np.array([np.size(X, 1), ])
+        cellarray = Cell2D(cols, rows, dtype=X.dtype)
+        x_cell = cellarray.mat2cell(X, cols, rows)
+        Y = []
+        for i in range(0, self.nblocks):
+            mat = self[i] @ x_cell[i]
+            Y.append(mat)
+        Y = np.concatenate(Y, axis=0)
+
+        if vector:
+            Y = Y[:, 0]
+
+        return Y
+
+
+def get_partition(blk_diag):
```

Comment:
Why is this necessary?

Reply:
Just for import syntax really. (The function itself, comes up in other notes.  Having both partition methods is vestigial from transitioning certain files using lists towards using the new class. I'll try to purge one).

```
from aspire.utils.blk_diag_matrix import get_partition
```
instead of either 
`BlkDiagononal.get_partition`
or people writing something like the following to keep their line length down.
`gp = BlkDiagononal.get_partition`

Reply:
In the process of implementing the collected suggestions, this code has been removed.